# ECeQ Project

## Introduction
The rapid development of high-throughput sequencing technology represents great opportunities for researchers to investigate genomic variations and transcriptomic changes. Multiple sequencing data are publicly available on multiple databases. But unfortunately, the analysis techniques are complexe and require multiple skills. The ECeQ provides the solution.

## About ECeQ : Shiny App for Gene Expression, CNV and eQTL data visualization
ECeQ project aims to create a biologist-friendly workflow to perform gene expression, genetic variation (CNV) and eQTL data analysis and  visualization.
The project includes a data analysis R package " AECeQ" and data visualization shiny app "VECeQ".

## AECeQ R package for data analysis
- 
- 

The AECeQ R package can be downloaded from  https://github.com/omicscodeathon/ECeQ/AECeQ_package. 

The guidelines for implementing this tool and related updates, are available at:  https://github.com/omicscodeathon/ECeQ/AECeQ/README.md.

## VECeQ Shiny App for data visualization
- 
- 

The VECeQ RShiny package can be downloaded from  https://github.com/omicscodeathon/ECeQ/VECeQ_package. 

The VECeQ Web shiny app  is publicly available at www.shinyapps.io/VECeQ.

The guidelines for implementing this tool and related updates, are available at:  https://github.com/omicscodeathon/ECeQ/VECeQ/README.md.

## Case study
To demonstrate the utility of the developed tool, we performed a study on the Brain-Blood-Barrier dysfunction in different neurological diseases ( HD, AD, ASD and  brain tumor) and in infectious diseases with neurological manifestation (SARS-CoV-2, TBEV, WNV, and Borrelia Garinii infection).

The analyzed data and results are available in : https://github.com/omicscodeathon/ECeQ/case_study.

## Abbreviations
- ECeQ : expression, CNV and eQTL analysis 

- AECeQ : Analyze expression, CNV and eQTL data

- VECeQ : Visualize expression, CNV and eQTL data

- DEGs : Differentially Expressed Genes

- CNV : Copy Number Variation

- eQTL : Expressed Quantitative Trait Loci

## Acknowledgements
This project was performed during the October 2022 Omics codeathon organized in collaboration with the African Society for Bioinformatics and Computational Biology (ASBCB). The authors thank the National Institutes of Health (NIH) Office of Data Science Strategy (ODSS) and the National Center for Biotechnology Information (NCBI) for their immense support before and during the codeathon.

## Contributors

- Hiba Ben Aribi, Faculty of Sciences of Tunis, University of Tunis El Manar (UTM), Tunis, Tunisia

- Imraan Dixon, Faculty of Health Sciences, University of Cape Town (UCT), Cape Town, South Africa 

- Kathryn Step, Faculty of Medicine and Health Sciences, Stellenbosch University (SU), Cape Town, South Africa
